python/triton_kernels/triton_kernels/tensor_details/layout_details/blackwell_scale.py

```diff
@@ -39,55 +39,223 @@ def __init__(self, shape) -> None:
+    in_base_rows  = tl.load(in_offsets_ptr  + pid_seg) # row offset in `data`
+    out_base_rows = tl.load(out_offsets_ptr + pid_seg) # row offset in `out`
+
+    # row indices (within this segment)
+    row_offsets = pid_row_blk * BLOCK_M + tl.arange(0, BLOCK_M)
+    col_offsets = tl.arange(0, BLOCK_N)
+
+    # global row indices in input/output
+    out_rows = out_base_rows + row_offsets[:, None]
+    out_cols = col_offsets[None, :]
+
+    in_rows = in_base_rows + row_offsets[:, None]
+    in_cols = col_offsets[None, :]
+
+    # masks
+    row_in_range_out = row_offsets < m_pad
+    col_in_range_out = col_offsets < K_pad
+    out_mask = row_in_range_out[:, None] & col_in_range_out[None, :]
+
+    row_in_range_in = row_offsets < m
+    col_in_range_in = col_offsets < K
+    in_mask = row_in_range_in[:, None] & col_in_range_in[None, :]
+
+    # default pad value = 0
+    # load from input where valid, else 0
+    vals = tl.zeros((BLOCK_M, BLOCK_N), dtype=tl.float32)
+
+    # compute linear ptrs with strides
+    in_ptrs = data_ptr + in_rows * stride_in_m + in_cols * stride_in_n
+    vals = tl.load(in_ptrs, mask=in_mask & out_mask, other=0.0)
+
+    # store into output
+    out_ptrs = out_ptr + out_rows * stride_out_m + out_cols * stride_out_n
+    tl.store(out_ptrs, vals, mask=out_mask)
+
+
+def pad_segments_triton(data, ex_hist, ex_hist_padded, K, K_pad, total_padded_rows):
+    """
+    data:           [total_M, K]       (contiguous or with known strides)
+    ex_hist:        [H] int32/long     (segment lengths)
+    ex_hist_padded: [H] int32/long     (padded lengths)
+    returns padded_data: [sum(ex_hist_padded), K_pad]
+    """
+    assert data.is_cuda
+    assert ex_hist.is_cuda and ex_hist_padded.is_cuda
+    assert ex_hist.dim() == 1 and ex_hist_padded.dim() == 1
+    assert ex_hist.shape == ex_hist_padded.shape
+
+    device = data.device
+    H = ex_hist.shape[0]
+
+    ex_hist_int = ex_hist.to(torch.int32)
+    ex_hist_padded_int = ex_hist_padded.to(torch.int32)
+
+    # prefix sums in *rows*
+    # in_offsets[i] = sum_{j < i} ex_hist[j]
+    # out_offsets[i] = sum_{j < i} ex_hist_padded[j]
+
+    # from exp_hist to offset format
+    in_offsets = torch.empty_like(ex_hist_int)
+    out_offsets = torch.empty_like(ex_hist_padded_int)
+
+    if H > 0:
+        in_offsets.zero_()
+        out_offsets.zero_()
+        if H > 1:
+            torch.cumsum(ex_hist_int[:-1], dim=0, out=in_offsets[1:])
+            torch.cumsum(ex_hist_padded_int[:-1], dim=0, out=out_offsets[1:])
+
+    # total_padded_rows = int(ex_hist_padded_int.sum().item())
+    padded_data = torch.empty(total_padded_rows, K_pad, device=data.device, dtype=data.dtype)
+
+    # strides (in elements, not bytes)
+    stride_in_m, stride_in_n = data.stride()
+    stride_out_m, stride_out_n = padded_data.stride()
+
+    BLOCK_M = 32
+    BLOCK_N = 64
+
+    grid = (
+        H,
+        triton.cdiv(128*2, BLOCK_M),
```

Comment:
This should be something like this vs. hardcoded
```suggestion
        triton.cdiv(triton.cdiv(total_padded_rows, H), BLOCK_M),
```

Reply:
I changed to use persistent kernel in the new implementation :)